eventstreamrpc/include/aws/eventstreamrpc/EventStreamClient.h

```diff
@@ -0,0 +1,659 @@
+             * is invoked, the `ClientConnection` is ready to be used for sending messages.
+             */
+            virtual void OnConnectCallback();
+            /**
+             * Invoked upon connection shutdown. `errorCode` will specify
+             * shutdown reason. A graceful connection close will set `errorCode` to
+             * `AWS_ERROR_SUCCESS` or 0.
+             */
+            virtual void OnDisconnectCallback(int errorCode);
+            /**
+             * Invoked upon receiving an error. Use the return value to determine
+             * whether or not to force the connection to close. Keep in mind that once
+             * closed, the `ClientConnection` can no longer send messages.
+             */
+            virtual bool OnErrorCallback(int errorCode);
+            /**
+             * Invoked upon receiving a ping from the server. The `headers` and `payload`
+             * refer to what is contained in the ping message.
+             */
+            virtual void OnPingCallback(
+                const Crt::List<EventStreamHeader> &headers,
+                const Crt::Optional<Crt::ByteBuf> &payload);
+        };
+
+        class AWS_EVENTSTREAMRPC_API ClientContinuationHandler
+        {
+          public:
+            /**
+             * Invoked when a message is received on this continuation.
+             */
+            virtual void OnContinuationMessage(
+                const Crt::List<EventStreamHeader> &headers,
+                const Crt::Optional<Crt::ByteBuf> &payload,
+                MessageType messageType,
+                uint32_t messageFlags) = 0;
+            /**
+             * Invoked when the continuation is closed.
+             *
+             * Once the continuation is closed, no more messages may be sent or received.
+             * The continuation is closed when a message is sent or received with
+             * the TERMINATE_STREAM flag, or when the connection shuts down.
+             */
+            virtual void OnContinuationClosed() = 0;
+        };
+
+        enum EventStreamRpcStatusCode
+        {
+            EVENT_STREAM_RPC_SUCCESS = 0,
+            EVENT_STREAM_RPC_NULL_PARAMETER,
+            EVENT_STREAM_RPC_INITIALIZATION_ERROR,
+            EVENT_STREAM_RPC_CONNECTION_CLOSED_BEFORE_CONNACK,
+            EVENT_STREAM_RPC_UNKNOWN_PROTOCOL_MESSAGE,
+            EVENT_STREAM_RPC_UNMAPPED_DATA,
+            EVENT_STREAM_RPC_UNSUPPORTED_CONTENT_TYPE,
+            EVENT_STREAM_RPC_STREAM_CLOSED_PREMATURELY,
+            EVENT_STREAM_RPC_CRT_ERROR
+        };
+
+        struct RpcError
+        {
+            EventStreamRpcStatusCode baseStatus;
+            int crtError;
+            operator bool() const noexcept { return baseStatus == EVENT_STREAM_RPC_SUCCESS; }
+        };
+
+        template <class T> class ProtectedPromise
+        {
+            /* A wrapper around std::promise so that it cannot be set twice without having to catch exceptions. */
```

Comment:
Sure, I can make it return a bool. Also, I actually try to avoid usage of this class unless absolutely necessary. However, some promises like the one used for Connect ACK or the initial response can be set twice because the connection or continuation can close at any time while data is being processed, so there's a potential race condition there. We could also bundle the mutex with the connection or continuation but that's slightly more repetitive.

Reply:
This race condition is the specific responsibility of the connection class. The goal is for it to be thread safe and manage it's own state correctly when it comes to concurrent operations.

Reply:
This is resolved in the latest iteration